### ISSUE TITLE
LXD grafana dashboard, wrong query in Loki panels

### ISSUE THREAD

**Issue by candlerb** [canonical/lxd#13165]
# Required information

 * Distribution: Ubuntu
 * Distribution version: 22.04
 * The output of "snap list --all lxd core20 core22 core24 snapd":

```
# snap list --all lxd core20 core22 core24 snapd
Name    Version       Rev    Tracking       Publisher   Notes
core20  20231123      2105   latest/stable  canonical✓  base,disabled
core20  20240111      2182   latest/stable  canonical✓  base
core22  20231123      1033   latest/stable  canonical✓  base,disabled
core22  20240111      1122   latest/stable  canonical✓  base
lxd     5.20-a8d6c52  26955  latest/stable  canonical✓  disabled,held
lxd     5.20-f3dd836  27049  latest/stable  canonical✓  held
```
(although note issue is with Grafana dashboard, not lxd)

# Issue description

This is regarding the Grafana dashboard at https://grafana.com/grafana/dashboards/19131-lxd/

> Aside: that page says "feedback welcome at https://discourse.ubuntu.com/c/lxd/126", but I was directed from [there](https://discourse.ubuntu.com/t/issue-report-for-lxd-loki-logging-in-grafana-dashboard/43363) to raise an issue here

I've set up Prometheus metrics and Loki logging, and have found an issue with the Grafana dashboard's handling of loki.

The Loki log panels at the bottom of the dashboard don't display anything. This is because the queries are wrong. They have `{app="lxd",type="lifecycle",instance="$job"}` for the first panel, and the second is the same but with `type="logging"`. However the "instance" label actually contains the hostname, not the job name or the container name.

I propose that at simplest, this needs to change to

```
{app="lxd",type="lifecycle",name=~"|$name"}
```

A more selective query is:

```
{app="lxd", type="logging", name=~"|$name"}  | logfmt | context_instance=~"|$name"
```

The vertical bar inside the regexp is because the "name" label may be missing. Here are some example logs:

```
2024-03-15T13:10:46Z {instance="nuc1", location="none", name="packetbeat", project="default", type="lifecycle"} action="instance-restarted" source="/1.0/instances/packetbeat" requester-address="@" requester-protocol="unix" requester-username="root" instance-restarted
2024-03-15T13:10:46Z {instance="nuc1", location="none", type="logging"}                                         context-action="shutdown" context-created="2022-06-24 12:38:48.192135561 +0000 UTC" context-ephemeral="false" context-instance="packetbeat" context-instanceType="container" context-project="default" context-timeout="10m0s" context-used="2023-09-07 08:38:24.929923055 +0000 UTC" level="info" Restarted instance
2024-03-15T13:10:39Z {instance="nuc1", location="none", type="logging"}                                         context-action="shutdown" context-created="2022-06-24 12:38:48.192135561 +0000 UTC" context-ephemeral="false" context-instance="packetbeat" context-instanceType="container" context-project="default" context-timeout="10m0s" context-used="2023-09-07 08:38:24.929923055 +0000 UTC" level="info" Restarting instance
2024-03-15T13:04:29Z {instance="nuc3", type="lifecycle"}                                                        action="config-updated" source="/1.0" requester-address="@" requester-protocol="unix" requester-username="root" config-updated
2024-03-15T13:04:25Z {instance="nuc2", type="lifecycle"}                                                        action="config-updated" source="/1.0" requester-address="@" requester-protocol="unix" requester-username="root" config-updated
2024-03-15T13:04:17Z {instance="nuc1", location="none", type="lifecycle"}                                       requester-protocol="unix" requester-username="root" action="config-updated" source="/1.0" requester-address="@" config-updated
2024/03/15 13:53:19 http://localhost:3100/loki/api/v1/query_range?direction=BACKWARD&end=1710507857767301302&limit=1000&query=%7Bapp%3D%22lxd%22%7D&start=1710507199542190712
2024/03/15 13:53:19 Common labels: {app="lxd", instance="nuc1", location="none", type="lifecycle"}
```

Notice how the first log has the name of the container as a label (`name="packetbeat"`), but some other logs relating to this container don't. They may have it buried in the logfmt data though, e.g. `context-instance="packetbeat"`

Hence the more sophisticated filter adds `| logfmt | context_instance=~"|$name"` (note that the hyphen is converted to underscore when loki parses the logfmt message to create a pseudo-label). If the name label is present, then it must contain the selected container instance name; similarly, if the log message contains context_instance, then it must match the selected container instance name.

# Steps to reproduce

1. Install loki
2. On lxd host, [turn on Loki logging](https://documentation.ubuntu.com/lxd/en/latest/howto/logs_loki/): `lxc config set loki.api.url=http://loki.example.net:3100`
3. Start and stop a container, check it creates some logs in Loki
4. Install grafana, add Loki as data source. Use "Explore" to browse Loki logs and check you can see the lxd logs.
5. Install the [lxd grafana dashboard](https://documentation.ubuntu.com/lxd/en/latest/howto/grafana/), 19131
6. Open the lxd dashboard, find that no Loki logs are visible
7. Edit panels and adjust the queries to those given above

# Information to attach

N/A

**Comment by nchorekchyan** [comment]
Hopefully you don't mind if I chip in, and this might already have been figured out:

As an FYI: The issue, at least in my case is as follows:

TLDR: You are using the wrong variable to filter the Loki logs. Add another variable for hosts/instances.

Long Version:

My Scenario: Multi-Node LXD Cluster. 
Prometheus has one job that collects from the cluster. Loki does not report by cluster, rather by server.

The dashboard has a variable $job, which corresponds to the prometheus **job_name field** configured in **prometheus's prometheus.yml** file, and is unique to prometheus. 

In my case, the variable dropdown for job shows lxd, as listed in the job_name of the prometheus.yml file. If I manually type in lxd-host-01, the prometheus side of the dashboard goes blank, but I get some logs in the Loki panels. (Note, the LXD cluster member names are host-01 and host02, not lxd-host-01 or lxd-host-02, which are the hostnames)

The variable $job is currently used for filtering the Loki panels in the expression

`"expr": "{app=\"lxd\", type=\"lifecycle\", instance=\"$job\"}",
`

Which means it will only work if the job name matches the LXD server hostname, and won't ever work in an LXD Cluster where multiple hosts need to report in.

None of the values Loki collects will match the Prometheus job name, so we need to pick something else.

**I was able to work around this by adding a separate variable, called $hosts, which is a simple query defined by the instance label (which in Loki's context means the hostname (I think) of the LXD cluster member, not to be confused with LXD cluster member name)**

So the updated Loki queries become (I switched from = to =~ to allow for multiple hosts to be selected)

`"expr": "{app=\"lxd\", type=\"lifecycle\", instance=~\"$hosts\"}",
`

I'm relatively new to Grafana but am familiar with Splunk, and in splunk there is an option to mix a query with a static selection list where you could set All to be a wildcard, I dont think that concept exists in grafana, so I can't quite figure out how to get All to work at the moment without rewriting a bunch, though single/multi host selection works.

This of course does not apply filtering to the prometheus based queries on the dashboard, but it's late and I just figured this out in a cram session and don't want to go through that exercise yet :-).

Hopefully this helps and/or is accurate.
